Refactor CropImagesModule

timothygebhard opened the pull request:
As discussed with @tomasstolker, I've slightly refactored the `CropImagesModule` such that it doesn't use the `apply_function_to_images()`-based parallelization anymore (which does not make a lot of sense here), but instead process the images in chunks and make use of numpy's vectorization features.

For testing purposes, I ran the module on a data set with 29681 frames. The old version took 960.2 seconds to crop the images, whereas the new version finished in just 29.8 seconds. This corresponds to a speed-up factor of ~32.

I also changed a few things in the `utils/image.py` in order to improve PEP8 compliance and make my linter stop complaining :)

@404NotFoundv2 I only realized when I was almost done with these changes that we had kinda assigned this task to you at last Wednesday's meeting 🙈I hope this is okay for you!

tomasstolker commented:
Excellent! Thanks a lot @timothygebhard!

- Completely removing the `copy` in `crop_image` was not possible? Because of how it was used by some other modules?

- I am not sure if we should replace the `elif` conditions with `else`. Although from a syntax perspective it makes indeed sense to use `else`, I find the `elif` better for readability of the code because it shows clearly the differences between the different cases. Perhaps add an `else` condition with an error instead?

timothygebhard commented:
> Completely removing the copy in crop_image was not possible? Because of how it was used by some other modules?

Exactly. Trying to remove the copying entirely causes the `test_waffle_center_odd` and `test_waffle_center_even` tests from `test_processing/test_centering.py` to fail, due to something that happens in the `WaffleCenteringModule`. 
If you know what's happening in there, feel free to change it so that we can drop the `copy` argument here, or at least change the default to False :)

> I am not sure if we should replace the elif conditions with else. Although from a syntax perspective it makes indeed sense to use else, I find the elif better for readability of the code because it shows clearly the differences between the different cases. Perhaps add an else condition with an error instead?

Agreed. I reverted this change and added a `RuntimeError` for the "impossible" fifth case that my linter was still complaining about ("local variable 'center' might be referenced before assignment").

tomasstolker commented:
> Exactly. Trying to remove the copying entirely causes the test_waffle_center_odd and test_waffle_center_even tests from test_processing/test_centering.py to fail, due to something that happens in the WaffleCenteringModule. If you know what's going on in there, feel free to change it so that we can drop the copy argument here, or at least change the default to False :)

Ok thanks, that is indeed worth looking into. I am not sure why the copy would be required for `WaffleCenterinModule`. Perhaps @18alex96 knows. This could be changed in a new PR.
